src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -0,0 +1,265 @@
+  }
+  public void deleteRecipe(String Id) {
+    DBUtils.blockOnFuture(DBUtils.recipe(Id).delete());
+    DBUtils.blockOnFuture(DBUtils.recipeMetadata(Id).delete());
+  }
+
+  public void editRecipeTitleContent(String Id, String editedTitle, String editedContent) {
+    DocumentReference contentRef = DBUtils.recipe(Id);
+    DocumentReference metadataRef = DBUtils.recipeMetadata(Id);
+    DBUtils.blockOnFuture(contentRef.update(Recipe.CONTENT_KEY, editedContent));
+    DBUtils.blockOnFuture(metadataRef.update(RecipeMetadata.TITLE_KEY, editedTitle));
+  }
+
+  public long voteRecipe(String Id, int voteDiff) {
+    DocumentReference metadataRef = DBUtils.recipeMetadata(Id);
+    long votes = DBUtils.blockOnFuture(metadataRef.get()).getLong(RecipeMetadata.VOTES_KEY);
+    metadataRef.update(RecipeMetadata.VOTES_KEY, votes + voteDiff);
+    return votes + voteDiff;
+  }
+
+  public List<RecipeMetadata> getAllRecipes(SortingMethod sortingMethod) {
+    Query recipesQuery = DBUtils.recipeMetadata();
+    switch (sortingMethod) {
+      case TOP:
+        recipesQuery = recipesQuery.orderBy(RecipeMetadata.VOTES_KEY, Query.Direction.DESCENDING);
+        break;
+      case NEW:
+        recipesQuery =
+            recipesQuery.orderBy(RecipeMetadata.TIMESTAMP_KEY, Query.Direction.DESCENDING);
+        break;
+    }
+    return DBUtils.blockOnFuture(recipesQuery.get()).toObjects(RecipeMetadata.class);
+  }
+
+  public List<Comment> getAllCommentsInRecipe(String recipeId, SortingMethod sortingMethod) {
+    Query commentsQuery = DBUtils.comments(recipeId);
+    switch (sortingMethod) {
+      case TOP:
+        commentsQuery = commentsQuery.orderBy(Comment.VOTES_KEY, Query.Direction.DESCENDING);
+        break;
+      case NEW:
+        commentsQuery = commentsQuery.orderBy(Comment.TIMESTAMP_KEY, Query.Direction.DESCENDING);
+        break;
+    }
+    return DBUtils.blockOnFuture(commentsQuery.get()).toObjects(Comment.class);
+  }
+
+  public List<Tag> getAllTags(boolean getHidden) {
+    Query tagsQuery =
+        getHidden ? DBUtils.tags() : DBUtils.tags().whereEqualTo(Tag.HIDDEN_KEY, false);
+    return DBUtils.blockOnFuture(tagsQuery.get()).toObjects(Tag.class);
+  }
+  
+  /** Returns a User object from userId. */
+    public User getUser(String userId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      DocumentSnapshot user = DBUtils.blockOnFuture(userRef.get());
+      return user.toObject(User.class);
+    }
+   
+    /** Adds a User to the db based on an input userId. */
+    public String addUser(String userId) {
+      DocumentReference newUserRef = DBUtils.users().document(userId);
+      User newUser = new User(userId);
+      DBUtils.blockOnFuture(newUserRef.set(newUser));  // .set() returns an ApiFuture
+      return userId;
+    }
+    
+    /** Delete a User from the db based on an input userId. */
+    public void deleteUser(String userId) {
+      DBUtils.blockOnFuture(DBUtils.user(userId).delete());
```

Comment:
should we also delete all recipe/comments from this user?

Reply:
Yes, that's a good point

Reply:
Actually, do we want to leave the recipes and comments created by a user but simply let the user displayed = [deleted]? We can just add a method like isUser(String uid) @bradleypickard @hari387

Reply:
Yeah that's a good call.